### PR TITLE
#24 add authorization for login

### PR DESCRIPTION
**Pull request by debboutr**
Be sure to update your `.env` file to add the url that you are running django at. I have noticed that finding the `sessionId` cookie is somewhat illusive in browsers so make sure that you have the `WHITELIST_CORS` set in your django app. It is now adding the cookie for me whether I use `http://127.0.0.1:8080` OR `http://localhost:8080` when viewing the app, but I have stopped/restarted so many times I'm not sure why it wasn't sometimes showing up when using `localhost`.

The only thing that I have added to the vuex store is the `username` to hold the value returned from django on a successful login, which gets dropped in any further bad response from the backend in the axios interceptor located in `src/main.js`. There is no way to get the cookie in js cause it is `HttpOnly`.

I have left in a button which is on the home page to ping the backend with a login that has bad credentials. This is the only way to see that our loggedIn state is valid because neither of the views that we have (home/about) use the backend so once the loggedIn state is set you will be able to access those forever. It is only when the backend is used that the logged in state will be checked. 

When you try the button on the home page, it will succeed until the cookie is bad and then you should be sent back to the `/login` page. You can set the `SESSION_COOKIE_AGE` in your django env to something small to test this out, unless you want to wait for 15 minutes. I've left some logging messages that you can checkout to see how it is working.

You can also go into the [VUE devtools extension](https://chrome.google.com/webstore/detail/vuejs-devtools/nhdogjmejiglipccpnnnanhbledajbpd) to alter the state back to not being logged in to see that you will then be sent to login as well.

UPDATE: In the tests I have added some listeners to attach the user to the `localStorage` on the window in order to mock the user being logged in so we don't have to go through the process of logging in with every cypress test moving forward.

### CODE REVIEW

File: src/views/Home.vue [hunk 1]
```diff
@@ -15,6 +20,36 @@ export default {
   components: {
     HelloWorld,
     KetcherWindow
+  },
+  methods: {
+    // REMOVE
```
Comment (mcovalt) [inline src/views/Home.vue]: Is this block supposed to be removed?

File: src/store/index.js [hunk 2]
```diff
@@ -3,9 +3,37 @@ import Vuex from "vuex";
...
+  mutations: {
+    SET_USER(state, user) {
+      state.user = user;
+    },
+    LOGOUT(state) {
+      state.user = "";
+    },
+    CYPRESS_SET_USER(state, user) {
+      state.user = user;
+    }
```
Comment (mcovalt) [inline src/store/index.js]: Why a separate `CYPRESS_SET_USER` mutation? Doesn't `SET_USER` do the same thing?

File: src/main.js [hunk 3]
```diff
@@ -27,22 +27,13 @@ const api = axios.create(axiosConfig);
...
-      });
-    }
-    throw new Error("Go back to start, do not pass go, do not collect $200!");
+  // this is currently our guard for loggedIn state
+  response => response,
+  error => {
+    store.dispatch("logout").then(() => {
+      router.push("/login");
+    });
+    return Promise.reject(error);
```
Comment (mcovalt) [inline src/main.js]: This will redirect all errors (even those unrelated to authentication) to the login page. Better to restrict to just error 401.

Comment (debboutr) [reply src/main.js]: I was unable to find a status_code on the error object, I have moved this to the beforeEach in the router.

**Pull request closed without merging**